### PR TITLE
Nh accommodate ramp pos error

### PR DESCRIPTION
**Pull request by noahheller**
This code has been rotting, so it does not follow best practices because the objective and logic keeps changing. If I had more time I would completely rewrite it however that is not the case

### CODE REVIEW

File: src/main/java/frc/robot/Robot.java [hunk 1]
```diff
@@ -73,6 +74,7 @@ public void disabledPeriodic() {
     public void autonomousInit() {
         ledEndgameTimer.restart();
         robotContainer.getLEDStrip().setPattern(RobotContainer.isRedAlliance() ? BlinkinPattern.HEARTBEAT_RED : BlinkinPattern.HEARTBEAT_BLUE);
+        new SetAutoRampPid(robotContainer.getRamp()).schedule();
```
Comment (bengold22) [inline src/main/java/frc/robot/Robot.java]: OOC why is this needed here instead of just setting in Ramp constructor?

File: src/main/java/frc/robot/commands/shooter/AdvancedSpinningShot.java [hunk 2]
```diff
@@ -54,10 +53,6 @@ public void execute() {
                 shooter.setShooterMotorLeftRPM(shooterSpeed.getLeftMotorSpeed());
             }
         }
```
Comment (bengold22) [inline src/main/java/frc/robot/commands/shooter/AdvancedSpinningShot.java:55]: No light strip color on Advanced Spinning Shot?

File: src/main/java/frc/robot/subsystems/LightStrip.java [hunk 3]
```diff
@@ -30,17 +30,22 @@ public void periodic() {
...
 
     public void setPattern(BlinkinPattern pattern) {
         this.pattern = pattern;
         colorSensorPort.set(pattern.getPwm());
-        SmartDashboard.putString("PATTERN",pattern.toString());
-        Logger.logDouble(baseLogName + "pwmSignal", pattern.getPwm(), Constants.ENABLE_LOGGING);
-        Logger.logString(baseLogName + "pwmName", pattern.toString(), Constants.ENABLE_LOGGING);
+        if (Constants.LED_DEBUG){
+            SmartDashboard.putString("PATTERN",pattern.toString());
+            Logger.logDouble(baseLogName + "pwmSignal", pattern.getPwm(), Constants.ENABLE_LOGGING);
```
Comment (bengold22) [inline src/main/java/frc/robot/subsystems/LightStrip.java:45]: I would vote for logging the signal whenever it's set, were these really that expensive? (leaving SmartDashboard  as debug makes sense)

File: src/main/java/frc/robot/utils/Alignable.java [hunk 4]
```diff
@@ -4,7 +4,7 @@
 
 public enum Alignable {
     AMP(0,0,0,0),
-    SPEAKER(-0.0381+0.234809,16.46,5.548,2.00);//1.343 - dist center of robot to speaker shot //value v2 1.343 - 0.447649 - Constants.ROBOT_LENGTH / 2
+    SPEAKER(-0.0381+0.234809,16.58-0.234809,5.548,2);//1.343 - dist center of robot to speaker shot //value v2 1.343 - 0.447649 - Constants.ROBOT_LENGTH / 2
```
Comment (bengold22) [inline src/main/java/frc/robot/utils/Alignable.java]: Is there a reason to leave these as math operations instead of just doing the math?

File: src/main/java/frc/robot/utils/TriFunction.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
+package frc.robot.utils;
```
Comment (bengold22) [inline src/main/java/frc/robot/utils/TriFunction.java:1]: Is this even used?

File: src/main/java/frc/robot/RobotContainer.java [hunk 6]
```diff
@@ -83,7 +83,7 @@ public class RobotContainer {
     private final Climber climber = new Climber();
     private final Vision vision = new Vision();
     private final IntakeSubsystem intake = new IntakeSubsystem();
-    private final PowerMonitor powerMonitor = new PowerMonitor();
+//    private final PowerMonitor powerMonitor = new PowerMonitor();
```
Comment (bengold22) [inline src/main/java/frc/robot/RobotContainer.java:86]: Should we at least leave a comment why this is commented out? Or just remove it if we don't need it.

File: src/main/java/frc/robot/commands/shooter/AdvancedSpinningShot.java (hunk 2 above)
Comment (bengold22) [reply src/main/java/frc/robot/commands/shooter/AdvancedSpinningShot.java:55]: OK so this is only used on the new joyLeftButton3.

Comment (noahheller) [reply src/main/java/frc/robot/commands/shooter/AdvancedSpinningShot.java:55]: would override auto ramp desiredPos indication

File: src/main/java/frc/robot/utils/Alignable.java (hunk 4 above)
Comment (noahheller) [reply src/main/java/frc/robot/utils/Alignable.java]: I was not confident on the right side of the equations. But the right side is known measured offset

File: src/main/java/frc/robot/utils/TriFunction.java (hunk 5 above)
Comment (noahheller) [reply src/main/java/frc/robot/utils/TriFunction.java:1]: it was used by now can be deleted

File: src/main/java/frc/robot/Robot.java [hunk 7]
```diff
@@ -73,6 +74,7 @@ public void disabledPeriodic() {
     public void autonomousInit() {
         ledEndgameTimer.restart();
         robotContainer.getLEDStrip().setPattern(RobotContainer.isRedAlliance() ? BlinkinPattern.HEARTBEAT_RED : BlinkinPattern.HEARTBEAT_BLUE);
+        CommandUtil.logged(new SetAutoRampPid(robotContainer.getRamp())).schedule();
```
Comment (armadilloz) [inline src/main/java/frc/robot/Robot.java:77]: Why do we need this?
If it is the case that the ramp PID is different than what it needs to be, then calling the subsystem synchronously here is better as it is done before any other command is scheduled and does not interrupt other ramp commands.

File: src/main/java/frc/robot/Robot.java [hunk 8]
```diff
@@ -92,6 +94,7 @@ public void teleopInit() {
         }
         CommandUtil.logged(new RaiseDeployer(robotContainer.getDeployer(), robotContainer.getLEDStrip())).schedule();
         CommandUtil.parallel("Reset Climber and Ramp", new teleOPinitReset(robotContainer.getRamp(), robotContainer.getClimber(), robotContainer.getLEDStrip())).schedule();
+        CommandUtil.logged(new SetTelopPid(robotContainer.getRamp())).schedule();
```
Comment (armadilloz) [inline src/main/java/frc/robot/Robot.java:97]: Same as the comment above

File: src/main/java/frc/robot/subsystems/LightStrip.java [hunk 9]
```diff
@@ -30,17 +30,22 @@ public void periodic() {
                 .ifPresent(c -> {
                     setPattern(predicateLightEvents.get(c));
```
Comment (armadilloz) [inline src/main/java/frc/robot/subsystems/LightStrip.java:31]: Maybe not for this PR, but if we're having memory issues, streams are not the best option.

File: src/main/java/frc/robot/subsystems/LightStrip.java (hunk 3 above)
Comment (armadilloz) [reply src/main/java/frc/robot/subsystems/LightStrip.java:45]: We were trying to save time to reduce the loop overruns so were moving every unnecessary code into debug. Can relax that for critical pieces, though. Command execution is one of them

File: src/main/java/frc/robot/utils/AutoAlignment.java [hunk 10]
```diff
@@ -22,30 +22,59 @@ public class AutoAlignment {
             Alignable.AMP, (x, y) -> new Rotation2d(Math.PI / 2),
             Alignable.SPEAKER, AutoAlignment::calcRobotRotationFaceSpeaker
     ));
-    private final static HashMap<Alignable, BiFunction<Double, Double, Rotation2d>> positionYawMap = new HashMap<>(Map.of(
-            Alignable.AMP, (x, y) -> Rotation2d.fromDegrees(Ramp.encoderToAngle(Constants.AMP_RAMP_ENC_VALUE)),
+    private final static HashMap<Alignable, BiFunction<Double, Translation3d, VelocityVector>> positionYawMap = new HashMap<>(Map.of(
+            Alignable.AMP, (aDouble, translation3d) -> new VelocityVector(10.5, Rotation2d.fromDegrees(Ramp.encoderToAngle(Constants.AMP_RAMP_ENC_VALUE))),
```
Comment (armadilloz) [inline src/main/java/frc/robot/utils/AutoAlignment.java:26]: Why 10.5? 
Didn't the AMP get removed?

File: src/main/java/frc/robot/utils/AutoAlignment.java [hunk 11]
```diff
@@ -22,30 +22,59 @@ public class AutoAlignment {
...
+     *
      * @param x the distance (unsigned) from the speaker on the xy plane
      * @param z the height between the top of the ramp and the middle of the speaker opening
      * @return a {@link Rotation2d} from the ground to the ramp representing the desired angle for shooting
+     * @see #calcRampAngle(double, double, double, double)
      */
-    private static Rotation2d calcRampAngle(double x, double z){
-        VelocityVector velocityVector = VectorUtils.fromVelAndDist(Constants.SHOOTER_VELOCITY, x, z, true);
+    @Deprecated
+    private static Rotation2d calcRampAngle(double x, double z, double vel) {
```
Comment (armadilloz) [inline src/main/java/frc/robot/utils/AutoAlignment.java:40]: This is now quite unreadable. There are 3 overloaded `calcRampAngle` and it is hard to determine which one is the one used in the lambda above (in line 27)

File: src/main/java/frc/robot/commands/ramp/RampFollow.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    public void end(boolean interrupted) {
+        lightStrip.setPattern(BlinkinPattern.BLACK);
+        ramp.setRampPos(Constants.RAMP_POS_STOW);
+    }
+
+    @Override
+    public void execute() {
+        Alignable alignableNow = drivetrain.getAlignable();
+        if (alignableNow != null) {
+            double diveTrainXVel = drivetrain.getFieldChassisSpeeds().vxMetersPerSecond;
```
Comment (armadilloz) [inline src/main/java/frc/robot/commands/ramp/RampFollow.java:49]:
```suggestion
            double driveTrainXVel = drivetrain.getFieldChassisSpeeds().vxMetersPerSecond;
```

File: src/main/java/frc/robot/commands/ramp/RampFollow.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+            double diveTrainYVel = drivetrain.getFieldChassisSpeeds().vyMetersPerSecond;
+            Translation3d rampPose = PoseUtils.addDimension(drivetrain.getPose().getTranslation(), Constants.ROBOT_FROM_GROUND);
+            double speakerRelativeXVel = diveTrainXVel * -1;
+            VelocityVector shooting = AutoAlignment.getYaw(alignable, rampPose, speakerRelativeXVel);
+            Translation3d futurePose = PoseUtils.getFieldEstimatedFuturePose(rampPose, diveTrainXVel, diveTrainYVel, 0.0, 0.05);
+            VelocityVector shootingFuture = AutoAlignment.getYaw(alignable, futurePose, speakerRelativeXVel);
+            if (shooting == null || shootingFuture == null) {
+                DriverStation.reportError("Invalid Odometry Can not shoot", true);
+                lightStrip.setPattern(BlinkinPattern.BLACK);
+                ramp.updateFF();
```
Comment (armadilloz) [inline src/main/java/frc/robot/commands/ramp/RampFollow.java:59]: What's this for?

File: src/main/java/frc/robot/commands/ramp/RampFollow.java [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+            if (Constants.RAMP_DEBUG){
+                SmartDashboard.putNumber("DEST DIFF", threshold);
+            }
+            Logger.logDouble("/robot/ramp/diffFromTarget", threshold, Constants.ENABLE_LOGGING);
+            if (isInThresh){
+                lightStrip.setPattern(BlinkinPattern.VIOLET);
+            } else {
+                lightStrip.setPattern(BlinkinPattern.BLACK);
+            }
+            ramp.updateFF();
```
Comment (armadilloz) [inline src/main/java/frc/robot/commands/ramp/RampFollow.java:93]: I did not review any of the calculations.
What's the updateFF for and why do we do that on every cycle?

File: src/main/java/frc/robot/commands/ramp/RampMoveAndWait.java [hunk 15]
```diff
@@ -43,6 +43,7 @@ public void initialize() {
 
     @Override
     public void execute() {
+        ramp.updateFF();
```
Comment (armadilloz) [inline src/main/java/frc/robot/commands/ramp/RampMoveAndWait.java:46]: Why?

File: src/main/java/frc/robot/subsystems/Ramp.java [hunk 16]
```diff
@@ -17,6 +17,8 @@ public class Ramp extends SubsystemBase {
     private final NeoPidMotor neoPidMotor;
     private double rampPos = Constants.RAMP_POS;
     private final InterpolatingDoubleTreeMap rampAngleMap = new InterpolatingDoubleTreeMap();
+    private double neoModerFF = Constants.RAMP_PID_FAR_FF;
```
Comment (armadilloz) [inline src/main/java/frc/robot/subsystems/Ramp.java:20]:
```suggestion
    private double neoModeFF = Constants.RAMP_PID_FAR_FF;
```

File: src/main/java/frc/robot/subsystems/Ramp.java [hunk 17]
```diff
@@ -32,15 +34,10 @@ private void configureMotor() {
         neoPidMotor.setSmartMotionAllowedClosedLoopError(Constants.RAMP_ERROR_RANGE);
         neoPidMotor.setMaxAccel(Constants.RAMP_MAX_RPM_ACCELERATION);
         neoPidMotor.getPidController().setP(Constants.RAMP_PID_P);
-        neoPidMotor.getPidController().setFF(Constants.RAMP_PID_FAR_FF);
+        neoPidMotor.getPidController().setFF(neoModerFF);
```
Comment (armadilloz) [inline src/main/java/frc/robot/subsystems/Ramp.java:37]: If this is being set in the `teleopInit` and `autoInit` then why do it in construction too?

File: src/main/java/frc/robot/subsystems/Ramp.java [hunk 18]
```diff
@@ -140,14 +137,40 @@ public boolean isShootAmpAngle(){
...
         //(y-b)/m=x
-        return (angle - 28.5) / 2.48;//needs be to measured again and put in constants
+        return (angle - Constants.RAMP_ENCODER_TO_ANGLE_Y_INTERCEPT) / Constants.RAMP_ENCODER_TO_ANGLE_SLOPE;
     }
 
     public void setAngle(Rotation2d angleFromGround) {
         setRampPos(angleToEncoder(angleFromGround.getDegrees()));
     }
 
+    public void setFF(double feedForward) {
```
Comment (armadilloz) [inline src/main/java/frc/robot/subsystems/Ramp.java:147]: Is this called externally too?

File: src/main/java/frc/robot/subsystems/Ramp.java [hunk 19]
```diff
@@ -140,14 +137,40 @@ public boolean isShootAmpAngle(){
...
+        } else if (getFF() != Constants.RAMP_PID_FAR_FF) {
+            setFF(Constants.RAMP_PID_FAR_FF);
+        }
+    }
+
     public double calcPose(Pose2d pose2d, Alignable alignable) {
         return rampAngleMap.get(pose2d.getTranslation().getDistance(new Translation2d(alignable.getX(), alignable.getY())));
     }
+
+    public void restorePid() {
```
Comment (armadilloz) [inline src/main/java/frc/robot/subsystems/Ramp.java:169]: probably should be called setDefaultFF() and setFarFF()

File: src/main/java/frc/robot/subsystems/Ramp.java [hunk 20]
```diff
@@ -140,14 +137,40 @@ public boolean isShootAmpAngle(){
...
 
+    public void setFF(double feedForward) {
+        neoPidMotor.getPidController().setFF(feedForward);
+        neoModerFF = feedForward;
+    }
+    public double getFF(){
+        return neoModerFF;
+    }
+
+    public void updateFF() {
```
Comment (armadilloz) [inline src/main/java/frc/robot/subsystems/Ramp.java:155]: Should this be called from some command execute or periodic?

File: src/main/java/frc/robot/utils/math/AngleUtils.java [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+     * Converts the angle to be between -PI/2 and PI/2.
+     * The function works by passing the sin of the angle to
+     * the arcSin function resulting in an angle clamped between -PI/2 and PI/2
+     * because of the range of the arcSin function
+     *
+     * @param angle to normalize
+     * @return normalized angle
+     */
+    public static Rotation2d normalize(Rotation2d angle) {
+        return new Rotation2d(Math.asin(Math.sin(angle.getRadians())));
```
Comment (armadilloz) [inline src/main/java/frc/robot/utils/math/AngleUtils.java:36]: Would this be faster by `reduce(angle+PI) - PI`?

**Review by armadilloz — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/Ramp.java (hunk 17 above)
Comment (armadilloz) [reply src/main/java/frc/robot/subsystems/Ramp.java:37]: Unknown. Leaving in place.,
